kube-runtime/src/controller/mod.rs

```diff
@@ -1085,6 +1089,89 @@ where
+    /// Usefull when there is a chain of ownership references, where the parent should trigger reconcile for all of the most distant children:
+    ///
+    /// ```
+    /// # async {
+    /// use futures::stream::StreamExt;
+    /// use k8s_openapi::api::core::v1::ConfigMap;
+    /// use k8s_openapi::api::apps::v1::Deployment;
+    /// use kube::{
+    ///     Client,
+    ///     api::{Api, ResourceExt},
+    ///     runtime::{
+    ///         reflector::ObjectRef,
+    ///         controller::{Controller, Action},
+    ///         watcher,
+    ///     },
+    /// };
+    /// use std::{convert::Infallible, io::BufRead, sync::Arc};
+    /// Controller::new(
+    ///     Api::<ConfigMap>::all(Client::try_default().await.unwrap()),
+    ///     watcher::Config::default(),
+    /// )
+    /// .watches_all(
+    ///     Api::<Deployment>::all(Client::try_default().await.unwrap()),
+    ///     watcher::Config::default(),
+    ///     |config_map, deployment| match deployment.annotations().get("config") {
+    ///         Some(name) if config_map.name_any() == name.clone() => Some(ObjectRef::from_obj(config_map)),
+    ///         _ => None,
+    ///     },
+    /// )
+    /// .run(
+    ///     |o, _| async move {
+    ///         println!("Reconciling {}", o.name_any());
+    ///         Ok(Action::await_change())
+    ///     },
+    ///     |_object: Arc<ConfigMap>, err: &Infallible, _| Err(err).unwrap(),
+    ///     Arc::new(()),
+    /// );
+    /// # };
+    /// ```
+    #[must_use]
+    pub fn watches_all<Other, I>(
+        mut self,
+        api: Api<Other>,
+        wc: watcher::Config,
+        mapper: impl Fn(&K, &Other) -> I + Sync + Send + Copy + 'static,
+    ) -> Self
+    where
+        Other: Clone + Resource + DeserializeOwned + Debug + Send + 'static,
+        Arc<Other>: Send,
+        I: IntoIterator<Item = ObjectRef<K>> + Sync + Send + 'static,
+        I::IntoIter: Sync + Send + 'static,
+    {
+        let store = self.store();
+        self.trigger_selector.push(
+            watcher(api, wc)
+                .touched_objects()
+                .filter_map(|o| future::ready(match o {
+                    Ok(o) => Some(o),
+                    Err(_) => None,
+                }))
+                .flat_map(move |o| {
+                    let other = Arc::new(o);
+                    stream::iter(store.state().into_iter().map(move |obj| {
+                        stream::iter(mapper(obj.as_ref(), other.as_ref()).into_iter().map(|r| {
+                            Ok(ReconcileRequest {
+                                obj_ref: r.clone(),
+                                reason: ReconcileReason::IndirectlyRelatedObjectUpdated {
+                                    obj_ref: r.to_string(),
+                                },
+                            })
+                        }))
+                    }))
+                })
```

Comment:
This now does a full store clone (and associated memory duplication) just to verify that the deployment __has__ a matching configmap.

Would it not be easier to fail early in the reconciler if we get a reconcile request for a deployment and we don't have the necessary configmap?

Reply:
Added another predicate for the cache lookup, so this exhaustive cloning may be avoided if needed.